File: java/0112-path-sum.java
class Solution {
    //This would be easily solved by DFS and then comparing the values
    public boolean dfs(TreeNode root, int targetSum, int currSum){
        if(root == null) return false;

        currSum += root.val;
        if(root.left == null && root.right == null){
            return (currSum == targetSum);
        }
        return dfs(root.left, targetSum, currSum) || dfs(root.right, targetSum, currSum);
    }
    public boolean hasPathSum(TreeNode root, int targetSum) {
        return dfs(root, targetSum, 0);
    }
}

File: kotlin/0112-path-sum.kt
class Solution {
    fun hasPathSum(root: TreeNode?, targetSum: Int): Boolean {
        
        root?: return false

        if(root.left == null && root.right == null) return targetSum - root.value == 0
        else return hasPathSum(root.left, targetSum - root.value) || hasPathSum(root.right, targetSum - root.value)
    }

    val TreeNode.value get() = this.`val`
}
